### ISSUE TITLE
Link an address to a contract as a library in deployer

### ISSUE THREAD

**Issue by bwheeler96** [trufflesuite/truffle#2838]
- [x] I've asked for help in the [Truffle Gitter](http://gitter.im/Consensys/truffle) before filing this issue.

---------------------------

## Issue

My contract has a `libary Lib {}` in it. I would like to use an existing library in the migrations script.

Something like:

`deployer.linkAddressAsLibrary('0xabcde...', 'AnExistingLibraryOnChain')`

**Comment by gnidan** [comment]
Looks like the missing piece is an equivalent for `deployer.deploy()`... right now the process is:

```javascript
deployer.deploy(MyLibrary);
deployer.link(MyLibrary, MyContract);
deployer.deploy(MyContract);
```

But you already have a deployed library... you just need that instance to be recorded in the artifact JSON. So I think a better solution is a deployer method to record the already-deployed library. Maybe something like this:

```javascript
deployer.recordDeployed(MyLibrary, "<address>");
deployer.link(MyLibrary, MyContract);
deployer.deploy(MyContract);
```

Something like that?

**Comment by bwheeler96** [comment]
@gnidan something like that yes. Here is the simplest use case I can think of:

`Library A` exists on main chain/kovan wherever, but for local development you want to use your own.

In local development, you link your own `Library A`, but then for mainnet migrations you use the on-chain version.

There are more complex cases, like what I am trying to do saving code by embedding constant addresses that don't belong in the code itself, but also don't belong in the constructor of the contract. So I use `library` as a way to embed constant addresses dynamically.

**Comment by bwheeler96** [comment]
@gnidan This actually already works, the key is that `.link` returns a thenable, so it must be awaited. Here is what works for me

```
await MyContract.link('CustomLibraryName', '0xabcde...')
```

Alternatively you could use `.then()` syntax.

Also, it should be noted that my deploy script must be setup like this for `async` to work:

```
async function doDeploy() {
  await ...
}

module.exports = (deployer, network) => {
  deployer.then(async () => {
    await doDeploy(deployer, network)
  })
}
```

**Comment by eggplantzzz** [comment]
Do note @bwheeler96 that the exported function in your migration can be `async`. So you can do
```
module.exports = async (deployer, network) => {
  const instance = await deployer.deploy(MyContract);
  ... // do other stuff
```

**Comment by bwheeler96** [comment]
@eggplantzzz I agree that this **should** be true, but someone, somewhere showed an example where there appeared to be some issue with exporting an async function.

I'm not sure if this is node version thing, not really a thing at all, or some other weirdness, but this code works for me and I'm afraid to change it :)

**Comment by eggplantzzz** [comment]
Oh, I have never had any problems making the exported migration functions async. If you come across a specific example where this doesn't work I would be very interested in seeing it to troubleshoot it!

**Comment by haltman-at** [comment]
Hm, I had been thinking `deployer.link(Library, Contract, libraryAddress)` for the syntax, but this approach possibly makes more sense, as my way might lead to a bunch of repetition.  We probably don't want to call it `linkAddressAsLibrary`, as it's not performing linking per se... `setAddressForLibrary`?  IDK, we can come up with something to call it.  Anyway I suppose any of these approaches would essentially solve the problem.

**Comment by cyrto** [comment]
It worked for me
```
MyLibrary.address = "<address>"
deployer.link(MyLibrary, MyContract);
deployer.deploy(MyContract);
```

**Comment by haltman-at** [comment]
That's an interesting hack that I imagine might work in some cases, but isn't really a general solution, and directly modifiying `MyLibrary.address` like that isn't really something we can endorse or support.

**Comment by gnidan** [comment]
> We could handle this by allowing the first argument of link to be a library instance rather than a library class. Then you could do await library = Library.at(address); deployer.link(library, Contract). That seems like it would be a fairly natural extension of the current syntax. What do people think of that?

That's great! Sorry we missed the suggestion @ayusht11!

**Comment by haltman-at** [comment]
Cool, I've put up #4752 to finally do something about this!

**Comment by ayusht11** [comment]
Thanks @gnidan @haltman-at

This would save me from a lot of trouble. :)

**Comment by lsqproduction** [comment]
Addressed in the release v5.5.2.
